doc/decisions/DR024-Versioning-traits-and-specifications-method.md

```diff
@@ -0,0 +1,412 @@
+> effectively make each version of a trait a new trait as far as the
+> `openassetio` core library is concerned.
+
+
+### Host side
+```python
+# The host side implementation is almost entirely unchanged from
+# from standard, excepting a new BatchElementError type
+trait_set = {LocatableContentTrait.kId}
+context = self.createTestContext()
+
+result = [None]
+
+def success_cb(idx, traits_data):
+    result[0] = traits_data
+
+def error_cb(idx, batchElementError):
+    if instanceOf(batchElementError, VersionNotSupportedError):
+      incompatible_with_manager_error()
+    ...
+
+self._manager.resolve(
+    [entity_reference], trait_set, ResolveAccess.kRead, context, success_cb, error_cb
+)
+```
+### Manager side
+```python
+# On the manager side, each trait gets unpacked (probably via a utility
+# method provided by traitgen or openassetio), and branching is done
+# on a trait by trait basis
+
+def resolve(
+        self,
+        entityReferences,
+        traitSet,
+        resolveAccess,
+        context,
+        hostSession,
+        successCallback,
+        errorCallback,
+    ):
+
+  # This function is imagined a pre-provided utility, possibly generated
+  # via traitgen, that would error in some way if the trait set does not
+  # contain the locatableContent trait.
+  locatableContent_version = locatableContent_version_from_trait_set(traitSet)
+
+  success_result = TraitsData()
+  if (locatableContent_version == 2):
+    trait = mediacreation.traits.content.LocatableContentTrait(success_result)
+    trait.setUrl(managed_filesystem_locations[ref.toString()])
+    successCallback(idx, success_result)
+  elif (locatableContent_version == 1):
+    trait = mediacreation.legacy.traits.content.LocatableContentTrait_v1(success_result)
+    trait.setLocation(managed_filesystem_locations[ref.toString()])
+    successCallback(idx, success_result)
+  else:
+    raise RuntimeError()
+
+```
+
+#### Pros
+- More flexible, traits of different versions can be mixed in the same
+  session, which could allow hosts/managers to perform partial upgrades
+  in the case they are blocked on small specifics, and allow components
+  of a system to each work with a different schema version and merge
+  their results.
+- Hosts don't have to interrogate and declare which library versions
+  are in use, which helps especially in the case of second-order
+  dependencies.
+- Seems to fit with existing OpenAssetIO design, especially with
+  `BatchElementError`, as an incompatible version wouldn't necessarily
+  need to error the whole batch.
+- No new API, core `openassetio` library remains version unaware.
+- Cannot be ignored, unlike other "layered on top" mechanisms.
+- Managers not required to consider storage/communication of additional
+  version data to enable interpretation by future downstream systems.
+- Prior art (e.g. OTIO) encodes versions in their data structure,
+  presumably for good reason.
+
```

Comment:
[fix] Add:
- Allows per-trait versioning, which avoids the need or "insider knowledge" as to which traits actually change in which package versions.
- Stateless
- Allows 'simple' managers and hosts to be written that don't understand specific data.

Reply:
[1] Again this seems like a criticism of a schema version rather than in-data versioning, but also again, there is some overlap so will add.
[2] Fine, but rephrasing because we both know it's not that simple.
[3] This restates the above point "Managers not required to consider storage/communication of additional
  version data to enable interpretation by future downstream systems."

Reply:
https://github.com/OpenAssetIO/OpenAssetIO/pull/1267/commits/61ba40856e4bda0344f7fcaeaac5c74a7602ac65

Reply:
> [3] This restates the above point "Managers not required to consider storage/communication of additional version data to enable interpretation by future downstream systems."

It was the **hosts** part in particular that didn't seem to be getting much air time in other points.

Reply:
That seems natural, hosts don't have the requirement to persist information on entities they don't understand, so don't run up against this problem.

Reply:
(for the record, as were in a good place on the DR anyway, but this is perhaps a useful long-term point). I was thinking more about the importance of being able to write a host that doesn't understand the data. Eg simpleResolver, a generic browser, systems that migrate assets from one system to another, how Gaffer would implement it etc... They're not storing the data, but they have no way to engage in a meaningful negotiation... if you see what I mean.

Reply:
Oh right, i guess that makes sense. Added it, although only about hosts and I think managers get called out on this area already in a slightly different way. https://github.com/OpenAssetIO/OpenAssetIO/pull/1267/commits/beed7f51c92087712fbb9dd42cf2e0690db65298